### PR TITLE
Support MSYS2 MinGW64 builds

### LINKED ISSUE #7: Include Windows build of V8
Hi,

I'd love to see V8 build be included in the deps.

Is this easily possible?

### LINKED ISSUE #44: build error on windowsx64
# rogchap.com/v8go
In file included from v8go.cc:3:
G:/����֪ʶѧϰ/golang/pkg/mod/rogchap.com/v8go@v0.2.0/deps/include/v8.h:7691:5: error: typedef 'v8::UnhandledExceptionCallback' is initialized (use decltype instead)
     _EXCEPTION_POINTERS* exception_pointers);
     ^~~~~~~~~~~~~~~~~~~
G:/����֪ʶѧϰ/golang/pkg/mod/rogchap.com/v8go@v0.2.0/deps/include/v8.h:7691:5: error: '_EXCEPTION_POINTERS' was not declared in this scope
G:/����֪ʶѧϰ/golang/pkg/mod/rogchap.com/v8go@v0.2.0/deps/include/v8.h:7691:26: error: 'exception_pointers' was not declared in this scope
     _EXCEPTION_POINTERS* exception_pointers);
                          ^~~~~~~~~~~~~~~~~~
G:/����֪ʶѧϰ/golang/pkg/mod/rogchap.com/v8go@v0.2.0/deps/include/v8.h:9462:7: error: 'UnhandledExceptionCallback' has not been declared
       UnhandledExceptionCallback unhandled_exception_callback);

### PR DESCRIPTION
**Pull request by neptoess**
These changes will allow v8go to work for users using the MSYS2 MinGW-w64 toolchain (https://www.msys2.org/) and the v8 package in their package manager,
`pacman -S mingw-w64-x86_64-v8`

Resolves #7, resolves #44

Note that the `snapshot_blob.bin` file will have to exist in your program's working directory. If the default install paths are used, this file is located at `C:\msys64\mingw64\bin\snapshot_blob.bin`

### CODE REVIEW

**Comment by mwm-tesla** [comment]
awesome!

**Comment by neptoess** [comment]
Resolves #44 #7

**Comment by rogchap** [comment]
Amazing @neptoess for getting this working on Windows.

One of the key design decisions of v8go is to include pre-built binaries so that users are able to simply use `go get` and it "just works".

If we're not able to do this for Windows we'll need to add detailed instructions to the README. Are you able to add this to the PR?

**Comment by neptoess** [comment]
@rogchap Adding prebuilt isn't a great option on Windows, in my opinion, because the default Go installer won't even leave the user with working cgo. Mingw-w64 gcc is the only compiler I'm aware of that allows creating native Windows binaries via cgo, and MSYS2 is a pretty popular way to acquire it, because of how awesome their pacman repo is. It's one extra command for the user to acquire libv8 in addition to gcc. Now, _if_ this package contained a static build of libv8, I would definitely say we should bundle it here. Because it doesn't, shipping a static libv8 for Windows means signing up to build said static library with the Mingw-w64 toolchain. I took a look at the scripts used to build the package (https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-v8/PKGBUILD) and decided that it would probably be significantly more effort to get a static library built than to just dynamically link V8.

tl;dr Yeah I'll modify the README :)

**Comment by neptoess** [comment]
@rogchap Let me know how this looks https://github.com/rogchap/v8go/pull/46/commits/458d4c7ef1f7927c44aae7f6a9b12f9d7a190bd1

**Comment by rogchap** [comment]
Yup; this is a fair compromise; might also be a better way to provide support for other architectures like freebsd etc. So static binaries for darwin and linux, but dynamic for the others 🤷‍♂️

File: README.md [hunk 1]
```diff
@@ -91,6 +91,18 @@ are included for Linux and OSX ie. you *should not* require to build V8 yourself
...
+While no prebuilt static V8 library is included for Windows, MSYS2 provides a package containing
+a dynamically linked V8 library that works.
+
+To set this up:
+1. Install MSYS2 (https://www.msys2.org/)
+2. Add the Mingw-w64 bin to your PATH environment variable (`C:\msys64\mingw64\bin` by default)
+3. Open MSYS2 MSYS and execute `pacman -S mingw-w64-x86_64-toolchain mingw-w64-x86_64-v8`
+4. This will allow building projects that depend on `v8go`, but, in order to actually run them,
+   you will need to copy the `snapshot_blob.bin` file from the Mingw-w64 bin folder to your program's
+   working directory (which is typically whereever `main.go` is)
```
Comment (rogchap) [inline README.md:104]:
```suggestion
   working directory (which is typically wherever `main.go` is)
```

Don't worry about fixing, I can update this later.

**Review by rogchap — APPROVED** [review]
Thank you. Looks good. Seems I need to fix the CI build (need to merge #6)

File: cgo.go [hunk 2]
```diff
@@ -1,7 +1,10 @@
 package v8go
 
-// #cgo CXXFLAGS: -fno-rtti -fpic -std=c++14 -I${SRCDIR}/deps/include
+// #cgo CXXFLAGS: -fno-rtti -fpic -std=c++14
+// #cgo !windows CXXFLAGS: -I${SRCDIR}/deps/include
+// #cgo windows CXXFLAGS: -DV8_COMPRESS_POINTERS -DV8_31BIT_SMIS_ON_64BIT_ARCH
```
Comment (rogchap) [inline cgo.go:5]: Looks like this is need for *nix too; will update once in master

Comment (neptoess) [reply cgo.go:5]: Are you referring to the pointer compression defines? If so, I can change CXXFLAGS to
```
// #cgo CXXFLAGS: -fno-rtti -fpic -std=c++14 -DV8_COMPRESS_POINTERS -DV8_31BIT_SMIS_ON_64BIT_ARCH
// #cgo !windows CXXFLAGS: -I${SRCDIR}/deps/include
```

Let me know if you want me to do this.

File: cgo.go [hunk 3]
```diff
@@ -1,7 +1,9 @@
 package v8go
 
-// #cgo CXXFLAGS: -fno-rtti -fpic -std=c++14 -I${SRCDIR}/deps/include -DV8_COMPRESS_POINTERS -DV8_31BIT_SMIS_ON_64BIT_ARCH
+// #cgo CXXFLAGS: -fno-rtti -fpic -std=c++14 -DV8_COMPRESS_POINTERS -DV8_31BIT_SMIS_ON_64BIT_ARCH
+// #cgo (darwin OR linux) CXXFLAGS: -I${SRCDIR}/deps/include
```
Comment (rogchap) [inline cgo.go:4]: Looks like this is not working as intended

**Review by rogchap — CHANGES_REQUESTED** [review]

**Comment by rogchap** [comment]
Hey @neptoess Thanks for continuing to update this PR. 
I've fixed up the test so I could check for any regressions in the PR.
The CI does not include any Windows tests; but we can add that suite later.

File: cgo.go (hunk 3 above)
Comment (neptoess) [reply cgo.go]: Fixed. I read this wrong :p https://golang.org/cmd/go/#hdr-Build_constraints

**Comment by neptoess** [comment]
> Hey @neptoess Thanks for continuing to update this PR.
> I've fixed up the test so I could check for any regressions in the PR.
> The CI does not include any Windows tests; but we can add that suite later.

I won't sign up to do it here, but I did this with Travis CI for confluent-kafka-go and it wasn't terrible
https://github.com/confluentinc/confluent-kafka-go/pull/564/files

**Review by neptoess — COMMENTED** [review]
I think I fixed it

**Review by rogchap — APPROVED** [review]
All green. Looking good. Thanks again @neptoess

**Comment by rogchap** [comment]
Published as part of [0.3.0](https://github.com/rogchap/v8go/releases/tag/v0.3.0)
